### PR TITLE
Bluetooth: Shell: Add a command cmd_send_traffic to send data traffic in coex scenario.

### LINKED ISSUE #67472: Add a command cmd_send_traffic to send data traffic in coex scenario.
**Is your feature request related to a problem? Please describe.**
The coex application we use bt-shell lacks console access when sending BLE traffic, resulting in us being unable to send wifi commands (wifi scan). The Root cause is that in the current code logic, the commands of ble and wifi are received through the shell_uart task, but ble always occupies the shell_uart task when sending data traffic, which causes the wifi command cannot be processed until the ble data traffic is completed.

**Describe the solution you'd like**
We have added a new command cmd_send_traffic and a timer(g_l2cap_send_timer), and then let ble send traffic process to be processed in this timer cb function(g_l2cap_send_timer_cb), so that ble will not always occupy the shell_task. When ble sends data traffic, the wifi command can also be Process normally.
image
image

**Describe alternatives you've considered**

**Additional context**
I'm making a PR to add this functionality, it will be submitted soon, please help with reviews and suggestions, thanks in advance!

### PR DESCRIPTION
**Pull request by nxf91057**
Added a new command cmd_send_traffic and a timer(g_l2cap_send_timer), and then let ble send data traffic process to be processed in this timer cb function(g_l2cap_send_timer_cb), so that ble will not always occupy the shell_task. When ble sends data traffic, the wifi command can also be Process normally.

Fixes #67472

Signed-off-by: Jasen Liu jasen.liu@nxp.com

### CODE REVIEW

**Comment by jhedberg** [comment]
@nxf91057 please remove the unrelated changes from this PR (i.e. everything else except for the changes to `shell/l2cap.c`). Also, you'll need to update the code to match Zephyr's coding style (in particular wrt to whitespace usage with indentation). Otherwise the CI Compliance check will fail.

**Comment by nxf91057** [comment]
@jhedberg Thank you for your comment. I have two pieces of information that i need to confirm with you.
1. Why should other files be removed? And only keep shell/l2cap.c  (The files i submitted are all useful for sending data traffic)
2. Do you mean that the zephyr code format cannot use spaces? Can i only use tabs? Are there any other format requirements?

**Comment by jhedberg** [comment]
> 1. Why should other files be removed? And only keep shell/l2cap.c  (The files i submitted are all useful for sending data traffic)

You did not explain the other changes anywhere (in particular *why* the changes are needed). They looked to me like something you had done for debugging purposes and just forgotten to remove before you created a pull request. At the very least these should be split into their own commits, each with a clear commit message that explains why the change is good or necessary. However, simply looking at these other changes, e.g. how they may substantially increase memory footprint or change the semantics of L2CAP flow control, I'm a bit skeptical that most (if any) will be acceptable. It's of course possible that I'm not fully understanding the background to some of these, and because of that it's critical that you provide good explanations & justifications for each change (at least in commit message, but code comments are also good whenever the code might not be immediately clear to the reader).

> 2. Do you mean that the zephyr code format cannot use spaces? Can i only use tabs? Are there any other format requirements?

See https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style

File: subsys/bluetooth/common/Kconfig [hunk 1]
```diff
@@ -96,7 +96,7 @@ config BT_BUF_ACL_RX_COUNT
 	int "Number of incoming ACL data buffers"
 	default NET_BUF_RX_COUNT if NET_L2_BT
 	default 3 if BT_RECV_BLOCKING
-	default 6
+	default 32
```
Comment (jori-nordic) [inline subsys/bluetooth/common/Kconfig:99]: this is a rather large increase for all applications. They probably don't use BT shell so they don't actually need that much.
Please set this in your app's `prj.conf` instead, and in `tests/bluetooth/shell/prj.conf`.

File: subsys/bluetooth/host/l2cap.c [hunk 2]
```diff
@@ -961,7 +961,8 @@ static void l2cap_chan_tx_init(struct bt_l2cap_le_chan *chan)
 	LOG_DBG("chan %p", chan);
 
 	(void)memset(&chan->tx, 0, sizeof(chan->tx));
-	atomic_set(&chan->tx.credits, 0);
+	/*In order to improve the packet sending speed of the sender, increase the initial credits value of tx to 30.*/
```
Comment (jori-nordic) [inline subsys/bluetooth/host/l2cap.c:964]: Please reformat the comments to fit zephyr style.

File: subsys/bluetooth/host/l2cap.c [hunk 3]
```diff
@@ -961,7 +961,8 @@ static void l2cap_chan_tx_init(struct bt_l2cap_le_chan *chan)
...
+	atomic_set(&chan->tx.credits, 30);
```
Comment (jori-nordic) [inline subsys/bluetooth/host/l2cap.c:965]: this is not legal. you can't just set an arbitrary number of credits on the peer's endpoint.

**Review by jori-nordic — CHANGES_REQUESTED** [review]
I just saw now the comments of @jhedberg . I agree fully with him. I will still publish my comments anyway.

This PR does way more than what it says both in the commit message and in the PR summary.

This area is also being refactored right now: see https://github.com/zephyrproject-rtos/zephyr/pull/67511

File: subsys/bluetooth/host/l2cap.c (hunk 3 above)
Comment (nxf91057) [reply subsys/bluetooth/host/l2cap.c:965]: hi jori, do you mean that tx and rx credits should be set the same? thanks!

Comment (jori-nordic) [reply subsys/bluetooth/host/l2cap.c:965]: no I mean you should not touch anything in `chan->tx`, as those params are set by the remote side.

Comment (jori-nordic) [reply subsys/bluetooth/host/l2cap.c:965]: and there's a test in the tree that uses it:
https://github.com/jori-nordic/zephyr/blob/9569e4410475fb62125e769a13222882902d8c75/tests/bsim/bluetooth/host/l2cap/credits_seg_recv/src/main.c#L423

**Comment by nxf91057** [comment]
Hi @jori-nordic Thanks for your comment. In response to the comment, i have updated the PR. Could you please help review it? Thanks!

**Comment by jori-nordic** [comment]
> In response to the comment, i have updated the PR

I still don't understand why you're touching `l2cap.c` at all. What is missing in the current API? You just want to send some traffic over l2cap, right?

**Comment by nxf91057** [comment]
Hi @jori-nordic With the current API, ble will always occupy the shell uart task when sending data traffic. If a wifi command is issued at this time, the command will not be processed, because the wifi command also needs to be received through the shell uart task.

**Comment by jori-nordic** [comment]
My point is you don't _have_ to mess with the core stack. You can just use the core stack API and re-implement an async shell_l2cap_send. It could use a worker thread (workqueue in zephyr). You can have as many workqueue threads as you wish and the priority is configurable.

**Comment by nxf91057** [comment]
Hi, @jori-nordic Regarding what you said about reimplementing asynchronous shell_l2cap_send, do you have relevant examples that you can share for reference? thanks!

**Comment by jori-nordic** [comment]
Please re-read my two comments above and the test it links to.
https://github.com/zephyrproject-rtos/zephyr/pull/67507#discussion_r1457304928

**Comment by nxf91057** [comment]
Hi @jori-nordici think we may have some mismatched information. Let me confirm with you again. Regarding what you said“I still don't understand why you're touching l2cap.c at all. What is missing in the current API? You just want to send some traffic over l2cap, right?”
1. I would like to ask which API you are talking about involving l2cap.c specifically refers to?
2. Some of the current APIs do not meet our needs, such as bt_l2cap_chan_give_credits, which is credits given to the peer device. What I want to achieve is to increase my initial credits value before sending data packets, which can increase the rate of sending packets. ->So i wrote an API：bt_l2cap_chan_give_credits_cb.
Looking forward for your reply, thank you！

**Pull request closed without merging**